File: typescipt/existing/src/language/types/tableType.ts





namespace Lexy.Compiler.Language.Types;

public class TableType : TypeWithMembers
{
   public string Type { get; }
   public Table Table { get; }

   public TableType(string type, Table table)
   {
     Type = type;
     Table = table;
   }

   protected bool Equals(TableType other)
   {
     return Type = other.Type;
   }

   public override bool Equals(object obj)
   {
     if (ReferenceEquals(null, obj)) return false;
     if (ReferenceEquals(this, obj)) return true;
     if (obj.GetType() ! GetType()) return false;
     return Equals((TableType)obj);
   }

   public override int GetHashCode()
   {
     return Type ! null ? Type.GetHashCode() : 0;
   }

   public override string ToString()
   {
     return Type;
   }

   public override VariableType MemberType(string name, IValidationContext context)
   {
     return name switch
     {
       "Count" => PrimitiveType.Number,
       Table.RowName => TableRowType(context),
       _ => null
     };
   }

   private TableRowType TableRowType(IValidationContext context)
   {
     var complexType = context.RootNodes.GetTable(Type)?.GetRowType(context);
     return new TableRowType(Type, complexType);
   }

   private IEnumerable<ComplexTypeMember> GetMembers(IValidationContext context)
   {
     return Table.Header.Columns.Select(column =>
         new ComplexTypeMember(column.Name, column.Type.CreateVariableType(context)))
       .ToList();
   }
}

File: src/Lexy.Compiler/Language/VariableTypes/TableType.cs
using System;
using Lexy.Compiler.Language.Functions;
using Lexy.Compiler.Language.Tables;
using Lexy.Compiler.Language.VariableTypes.Functions;

namespace Lexy.Compiler.Language.VariableTypes;

public class TableType : TypeWithMembers
{
    public string TableName { get; }
    public Table Table { get; }

    public TableType(string tableName, Table table)
    {
        TableName = tableName;
        Table = table;
    }

    protected bool Equals(TableType other)
    {
        return TableName == other.TableName;
    }

    public override bool Equals(object obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != GetType()) return false;
        return Equals((TableType)obj);
    }

    public override int GetHashCode()
    {
        return TableName != null ? TableName.GetHashCode() : 0;
    }

    public override string ToString()
    {
        return TableName;
    }

    public override VariableType MemberType(string name, IComponentNodeList componentNodes)
    {
        switch(name)
        {
            case Table.CountName:
                return PrimitiveType.Number;
            case Table.RowName:
                return TableRowType(componentNodes);
        }

        return Table.Header?.GetColumn(name) != null
            ? new GeneratedType(name, Table, GeneratedTypeSource.TableColumn, Array.Empty<GeneratedTypeMember>())
            : null;
    }

    public override IInstanceFunction GetFunction(string name)
    {
        return name switch
        {
            LookUpFunction.Name => new LookUpFunction(Table),
            LookUpRowFunction.Name => new LookUpRowFunction(Table),
            _ => null
        };
    }

    private GeneratedType TableRowType(IComponentNodeList componentNodes)
    {
        return componentNodes.GetTable(TableName)?.GetRowType();
    }
}